api/v1alpha4/virtualmachine_affinity_types.go

```diff
@@ -0,0 +1,232 @@
+	// set to topology.kubernetes.io/zone, it means this term applies to nodes
+	// with that label present.
+	TopologyKey string `json:"topologyKey"`
+}
+
+// WeightedVMAffinityTerm defines the weighted VM affinity/anti-affinity term.
+type WeightedVMAffinityTerm struct {
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=100
+
+	// Weight describes the weight of the associated term.
+	Weight int32 `json:"weight"`
+
+	// VMAffinityTerm describes the affinity term.
+	VMAffinityTerm VMAffinityTerm `json:"vmAffinityTerm"`
+}
+
+// VirtualMachineAffinityVMAffinitySpec defines the affinity requirements for scheduling
+// rules related to other VMs.
+type VirtualMachineAffinityVMAffinitySpec struct {
+	// +optional
+	// +listType=atomic
+
+	// RequiredDuringSchedulingIgnoredDuringExecution describes affinity
+	// requirements that must be met or the VM will not be scheduled.
+	//
+	// When there are multiple elements, the lists of nodes corresponding to
+	// each term are intersected, i.e. all terms must be satisfied.
+	RequiredDuringSchedulingIgnoredDuringExecution []VMAffinityTerm `json:"requiredDuringSchedulingIgnoredDuringExecution,omitempty"`
+
+	// +optional
+	// +listType=atomic
+
+	// PreferredDuringSchedulingIgnoredDuringExecution describes affinity
+	// requirements that should be met, but the VM can still be scheduled if
+	// the requirement cannot be satisfied. The scheduler will prefer to schedule VMs to nodes
+	// that satisfy the anti-affinity expressions specified by this field, but it may choose a node that
+	// violates one or more of the expressions. The node that is most preferred is the one with the
+	// greatest sum of weights.
+	//
+	// When there are multiple elements, the lists of nodes corresponding to
+	// each term are intersected, i.e. all terms must be satisfied.
+	PreferredDuringSchedulingIgnoredDuringExecution []WeightedVMAffinityTerm `json:"preferredDuringSchedulingIgnoredDuringExecution,omitempty"`
+}
+
+// VirtualMachineAntiAffinityVMAffinitySpec defines the anti-affinity requirements for scheduling
+// rules related to other VMs.
+type VirtualMachineAntiAffinityVMAffinitySpec struct {
+	// +optional
+	// +listType=atomic
+
+	// RequiredDuringSchedulingIgnoredDuringExecution describes anti-affinity
+	// requirements that must be met or the VM will not be scheduled.
+	//
+	// When there are multiple elements, the lists of nodes corresponding to
+	// each term are intersected, i.e. all terms must be satisfied.
+	RequiredDuringSchedulingIgnoredDuringExecution []VMAffinityTerm `json:"requiredDuringSchedulingIgnoredDuringExecution,omitempty"`
+
+	// +optional
+	// +listType=atomic
+
+	// PreferredDuringSchedulingIgnoredDuringExecution describes anti-affinity
+	// requirements that should be met, but the VM can still be scheduled if
+	// the requirement cannot be satisfied. The scheduler will prefer to schedule VMs to nodes
```

Comment:
Please change `schedule VMs to nodes` to `schedule VMs`. The scheduled location could be a node, a zone, etc. Let's keep it ambiguous here.